File: framework/scala/src/main/scala/soda/scala/unittest/conv/ObjectConverter.scala
package soda.scala.unittest.conv

import play.api.libs.json.JsValue

trait ObjConv {
  def as[E]: ObjectConverter[E] = this.asInstanceOf[ObjectConverter[E]]
}

trait ObjectConverter[E] extends ObjConv {

  def fromJsonSerializable(js: JsValue): E

  def toJsonSerializable(element: E): JsValue

}

File: framework/php/src/Soda/Unittest/ObjectConverter.php
<?php
namespace Soda\Unittest;

class ObjectConverter
{
    private mixed $parser;
    private mixed $serializer;

    function __construct($p, $s)
    {
        $this->parser = $p;
        $this->serializer = $s;
    }

    function fromJsonSerializable($js)
    {
        return ($this->parser)($js);
    }

    function toJsonSerializable($obj)
    {
        return ($this->serializer)($obj);
    }
}